dropshot/src/server.rs

```diff
@@ -257,17 +252,41 @@ impl<C: ServerContext> InnerHttpServerStarter<C> {
     /// Begins execution of the underlying Http server.
     fn start(
         self,
-        close_signal: tokio::sync::oneshot::Receiver<()>,
+        mut close_signal: tokio::sync::oneshot::Receiver<()>,
         log_close: Logger,
-    ) -> tokio::task::JoinHandle<Result<(), hyper::Error>> {
-        let graceful = self.0.with_graceful_shutdown(async move {
-            close_signal.await.expect(
-                "dropshot server shutting down without invoking close()",
-            );
-            info!(log_close, "received request to begin graceful shutdown");
-        });
+    ) -> tokio::task::JoinHandle<()> {
+        use hyper_util::rt::{TokioExecutor, TokioIo, TokioTimer};
+        use hyper_util::server::conn::auto;
+
+        tokio::spawn(async move {
+            let mut builder = auto::Builder::new(TokioExecutor::new());
+            // http/1 settings
+            builder.http1().timer(TokioTimer::new());
+
+            let graceful =
+                hyper_util::server::graceful::GracefulShutdown::new();
 
-        tokio::spawn(graceful)
+            loop {
+                tokio::select! {
+                    Ok((sock, remote_addr)) = self.0.accept() => {
```

Comment:
What if `accept()` returns an error?  The [docs](https://docs.rs/tokio/1.38.0/tokio/net/struct.TcpListener.html) aren't super explicit but I gather this could happen any time the underlying accept(2) syscall returns an error, which can include transient errors like ECONNABORTED (on illumos, anyway).  EMFILE might also be considered transient.  (The docs also say "These would terminate the stream if not handled in any way."  I'm not sure what that means.  This struct is not a `Stream` and I don't know how it could know whether you handled the error.)

If this happens, according to the [`tokio::select!`](https://docs.rs/tokio/1.38.0/tokio/macro.select.html) docs:

> Once an <async expression> returns a value, attempt to apply the value to the provided <pattern>, if the pattern matches, evaluate <handler> and return. If the pattern does not match, disable the current branch and for the remainder of the current call to select!. Continue from step 3.

Step 3 waits for another of the futures to complete.  So my interpretation of all of this is: if we encounter a transient error from `accept()`, this program will stop accepting new connections.  This would be very bad.

Besides fixing this, it makes me a little nervous that we're implementing something so delicate and hard-to-test here.  Is there some existing implementation we could use here?  Would it help to turn this into a https://docs.rs/tokio-stream/0.1.15/tokio_stream/wrappers/struct.TcpListenerStream.html?

Reply:
Good catch, hyper did handle all those before. But as part of the removal of the `Accept` trait, due to all of it's subtle issues, that was removed as well. Something generic _might_ appear again in hyper-util, and when it does, Dropshot could use it.

For now, I can inline the behavior that used to exist.

Reply:
I've added an `HttpAcceptor` type here that does what hyper used to do.